sys/net/credman/credman.c

```diff
@@ -90,6 +100,246 @@ int credman_add(const credman_credential_t *credential)
+        DEBUG("credman: the private key information should be contained in an ASN.1 SEQUENCE\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to version */
+    asn1_tree *node = priv_key.child;
+
+    if (!node || node->type != ASN1_TYPE_INTEGER) {
+        DEBUG("credman: invalid private key version\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKeyAlgorithm */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_SEQUENCE || !node->length) {
+        DEBUG("credman: invalid private key algorithm identifier\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* for now only ECDSA is supported by credman */
+    /* the algorithm should be Elliptic Curve Public Key (OID 1.2.840.10045.2.1) */
+    asn1_tree *algorithm = node->child;
+    if (sizeof(ecPublicKey) != algorithm->length) {
+        DEBUG("credman: wrong private key algorithm, only ecPublicKey is supported\n");
+        return CREDMAN_INVALID;
+    }
+
+    if (memcmp(ecPublicKey, algorithm->data, sizeof(ecPublicKey)) != 0) {
+        DEBUG("credman: wrong OID for algorithm. Expected 1.2.840.10045.2.1\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKey */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_OCTET_STRING || !node->data || !node->length) {
+        DEBUG("credman: no private key found\n");
+        return CREDMAN_INVALID;
+    }
+
+    return credman_load_private_ecc_key(node->data, node->length, cred);
+
+}
+
+int credman_load_private_ecc_key(const uint8_t *buf, size_t buf_len, credman_credential_t *cred)
+{
+    assert(buf);
+    assert(cred);
+
+    int obj_count = der_object_count((uint8_t *)buf, buf_len);
+    if (obj_count < 0) {
+        DEBUG("credman: could not calculate the number of elements within the key\n");
+        return CREDMAN_INVALID;
+    }
+
+    asn1_tree objects[obj_count];
+    asn1_tree priv_key;
+
+    if (der_decode((uint8_t *)buf, buf_len, &priv_key, objects, obj_count) < 0) {
+        DEBUG("credman: could not parse the key\n");
+        return CREDMAN_INVALID;
+    }
+
+    /*
+     * From https://tools.ietf.org/html/rfc5915#section-3
+     *
+     * ECPrivateKey ::= SEQUENCE {
+     *      version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+     *      privateKey     OCTET STRING,
+     *      parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
+     *      publicKey  [1] BIT STRING OPTIONAL
+     * }
+     */
+
+    /* point to version, it SHALL be 1 */
+    asn1_tree *node = priv_key.child;
+    if (!node || node->type != ASN1_TYPE_INTEGER || node->data[0] != 0x01) {
+        DEBUG("credman: invalid private key version\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKey */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_OCTET_STRING || !node->data || !node->length) {
+        DEBUG("credman: invalid private key\n");
+        return CREDMAN_INVALID;
+    }
+
+    cred->type = CREDMAN_TYPE_ECDSA;
```

Comment:
How do you know that the private key here is ECC if the user calls `credman_load_private_ecc_key` directly? I don't see the same check as [[1]] in this function.

[1]: https://github.com/RIOT-OS/RIOT/pull/16263/files#diff-1d82b7a03a2ca314bc6e3fe4ea40dd084f18086291491faa17152836d9d1f40eR234-R237

Reply:
As you can see in the commented lines above, there is no OID identifier in the `ECPrivateKey` structure, so no way to check this. The user needs to know that the key is ECC, we can at most check for a version and that the structure follows the expected format, which is done.

Reply:
I'm thinking if we should add a member to the EC keys specifying the used curve

Reply:
Thinking from the side of integrating with other (D)TLS libraries, I think it makes sense to have it along with the keys. Then when loading the key, we can use it to check if the currently used library supports the curve. For example here, tinydtls only support one curve but wolfssl supports [multiple curves][1]. Bonus point, this is also how it's done in [BearSSL][2].

[1]: https://www.wolfssl.com/using-supported-elliptic-curves-extension-with-wolfssl/
[2]: https://www.bearssl.org/gitweb/?p=BearSSL;a=blob;f=inc/bearssl_ec.h;h=acd3a2bf5a5550fe99130d9e4b99f7db2169c69d;hb=HEAD#l248